packages/@aws-cdk/core/lib/stack.ts

```diff
@@ -848,12 +848,32 @@ function merge(template: any, part: any) {
     if (!dest) {
       template[section] = dest = src;
     } else {
-      // add all entities from source section to destination section
-      for (const id of Object.keys(src)) {
-        if (id in dest) {
-          throw new Error(`section '${section}' already contains '${id}'`);
+      if (typeof dest === 'object') {
```

Comment:
This will do in a pinch, but you're assuming the source templates already conform to a schema without mistakes, and your code will do unexpected things otherwise.  Also, this code is getting long in the tooth and needs to be broken up into functions which each do a clearly defined small thing.

I'd rather see something like this:

```ts
function merge(a: object, b: object) {
  const ret: any = {};
  for (const key of new Set(Object.keys(a).concat(Object.keys(b)))) {
    ret[key] = mergeSection(key, a[key], b[key]);
  }
  return ret;
}

function mergeSection(section: string, a: any, b: any) {
  switch (section) {
    case 'Description': /* ... validate a and b are string | undefined */ return `${a}\n${b}`;
    case 'AWSTemplateFormatVersion': // ...;
    case 'Resources':
    case 'Conditions':
    case 'Parameters':
    case 'Outputs':
     return mergeObjectsWithoutDuplicates(a, b);  // mergeObjects should validate that they are indeed objects | undefined
    default: 
     throw new Error(`Wut? I don't know how to merge ${section}`);
  }
}
```

Reply:
Done!